lib/rex/proto/http/proxy.rb

```diff
@@ -0,0 +1,492 @@
+    sub_response_headers(resp,headers)
+  end
+
+  #
+  # Sends a 404 error to the client for a given request.
+  #
+  def send_e404(cli, request)
+    resp = Response::E404.new
+
+    resp['Content-Type'] = 'text/html'
+
+    resp.body =
+      "<html><head>" +
+      "<title>404 Not Found</title>" +
+      "</head><body>" +
+      "<h1>Not found</h1>" +
+      "The requested URL #{html_escape(request.resource)} was not found on this server.<p><hr>" +
+      "</body></html>"
+
+    # Send the response to the client like what
+    cli.send_response(resp)
+  end
+
+  # Rewrite proxy elements of request to target request
+  # Determine whether or not to use SSL for outbound
+  def proxy_header_rewrite(request)
+    if request.uri.match(/^http/i)
+
+      # Determine if we need to use SSL
+      request.headers['SSL'] = (request.uri[4].downcase == 's')
+      # Rewrite URI HTTP request to headers and proper URI
+      proxy_keys = request.headers.select {|key,val| key.match(/^proxy/i)}
+      # Remove the proxy header info
+      request.headers - proxy_keys
+      # Rewrite the proxy specific options to our direct headers
+      proxy_keys.each do |k,v|
+        request.headers[k.sub(/^proxy-/i,'')] = v
+      end
+      normalize_request_target(request)
+    end
+  end
+
+  def normalize_request_target(request)
+    # Extract host port and URI, map as appropriate, account for emtpy
+    target = request.uri[/\/([\w\d\.]+)/, 1]
+
+    host,port = target.split(':')
+    targ_uri = request.uri[/\/[\w\d\.]+(\/.*)/,1]
+    request.uri = targ_uri || '/'
+
+    # Check for DNS hostname
+    if host.match(/\w+\.\w+/)
+      request.headers['Vhost'] = host
+      host = Rex::Socket.addr_itoa(
+       Rex::Socket.gethostbyname( host )[3].unpack( 'N' ).first
+      )
+      request.headers['Host'] = port.nil? ? host : "#{host}:#{port}"
+    else
+      # Try an IPv6 configuration if host is not an IPv4 addr
+      if !Rex::Socket.is_ipv4?(host)
+        host = request.headers['Host'].reverse.split(':',2).last.reverse
+        if !Rex::socket.is_ipv6?(host)
+          raise ::Rex::ConnectionError('Cannot determine request target')
+        end
+      end
+      # If we're here our host is valid, write the headers
+      request.headers['Host'] = port.nil? ? host : "#{host}:#{port}"
+    end
+  end
+
+  # Determine if redirect is valid
+  def valid_redirect?(response)
+    return ([301,302].any? {response.code} and response.headers['Location'])
+  end
+
+  attr_accessor :listen_port, :listen_host, :context, :ssl, :comm, :ssl_cert
+  attr_accessor :listener, :proxies, :clients, :req_handler, :res_handler
+  attr_accessor :redirect_limit, :rewrite_proxy_headers, :request_timeout
+  attr_accessor :connect_host, :connect_port
+
+protected
+
+  #
+  # Watches client connections for data, forwards back to client
+  #
+  def monitor_clients
+    self.clients.each do |client|
+      if client.session and resp = client.session.read_response(2)
```

Comment:
Is this meant to be `=` assignment or `==` comparison?

In the case of the former, please keep variable assignment outside of conditional statements.

Reply:
Variable assignment outside conditional statements? The idea is resp can't be assigned if the session is nil. I suppose I can be more explicit about that, but out of curiosity, why don't we allow this pattern? If unwound in full notation, part after and is another block with resp assignment scoped in it and no calls to it if the condition fails. I could see a change which permits ref to resp when its unassigned causing a "don't call methods on nil" problem...
Got a preferred prototype for this sort of deterministic assignment, or just a "bunny slope" if...then...end?

Reply:
I have no idea whether the pattern is "allowed" but something like this might make more sense:

```ruby
      resp = nil
      if client.session
        resp = client.session.read_response(2)
        client.send_response(resp) if resp
      end
```

What's wrong with this code?

```c
if ((options == (__WCLONE|__WALL)) && (current->uid = 0))
        retval = -EINVAL;
```

Reply:
Ruby is evaluated, not compiled, so that's not exactly analogous. I've not tried that trick in c, actually not sure it would work at all since youre evaluating an immediate assignment, but I imagine your original intent was == not =.

So much for me trying to be clever. Will review tonight. :)

Reply:
Also, your code is broken for ruby. My code deals with client being nil, yours will raise when it is.

Reply:
It looks like [best practice](https://github.com/bbatsov/ruby-style-guide#safe-assignment-in-condition) is to wrap variable assignment within conditional statements in brackets to indicate that it's an intentional assignment and not an accidental typo in an equality comparison, like so:

```ruby
if client.session && (resp = client.session.read_response(2))
```